src/main/scala/io/iohk/ethereum/blockchain/sync/PeersClient.scala

```diff
@@ -209,18 +209,23 @@ object PeersClient {
   }
 
   //returns the next best peer after the one already returned previously
-  //TODO: make sure the next best peer has a different best block, so we don't fetch in parallel identical branches
   //TODO: whenever this method is called - do activeFetchingNodes.add(_) on the peer returned
-  def nextBestPeer(peersToDownloadFrom: Map[PeerId, PeerWithInfo]): Option[Peer] = {
+  def nextBestPeer(
+      peersToDownloadFrom: Map[PeerId, PeerWithInfo],
+      activeFetchingNodes: Set[PeerWithInfo]
+  ): Option[Peer] = {
     val peersToUse = peersToDownloadFrom.values
-      .collect { case PeerWithInfo(peer, PeerInfo(_, chainWeight, true, _, _)) =>
-        (peer, chainWeight)
+      .collect { case PeerWithInfo(peer, peerInfo @ PeerInfo(_, _, true, _, _)) =>
+        (peer, peerInfo)
       }
 
     val peer =
       peersToUse
-        .filter { case (peer, _) => !activeFetchingNodes.contains(peer) }
-        .maxByOption { case (_, weight) => weight }
+        .filterNot { case (peer, _) => activeFetchingNodes.map(_.peer).contains(peer) }
+        .filterNot { case (_, peerInfo) =>
+          activeFetchingNodes.map(_.peerInfo).exists(PeerComparator.doPeersHaveSameBestBlock(peerInfo, _))
```

Comment:
In theory, this line implies the same as the line above, right? Cause a peer will have the same best block as itself?

Reply:
in theory the peers are progressing forward. So if we added the peer to the activeFetchingNodes 10min ago, the peersToDownloadFrom can already contain same peer with updated best block

Reply:
Ahh ok, that's fun... Immutable objects in a mutable datastructure.

Reply:
at this point, the data structure is already immutable :)